samples/apps/cap/py/src/ag_adapter/AG2CAP.py

```diff
@@ -0,0 +1,76 @@
+from ag_adapter.AutoGenConnector import AutoGenConnector
+from autogen import Agent, ConversableAgent
+import time
+from typing import Callable, Dict, List, Optional, Union
+
+
+class AG2CAP(ConversableAgent):
+    """
+    A conversable agent proxy that sends messages to CAN when called
+    """
+
+    def __init__(
+        self,
+        network,
+        agent_name: str,
+        agent_description: Optional[str] = None,
+    ):
+        super().__init__(name=agent_name, description=agent_description)
+        self._agent_connector: AutoGenConnector = None
+        self._network = network
+        self._recv_called = False
+
+    def reset_receive_called(self):
+        self._recv_called = False
+
+    def was_receive_called(self):
+        return self._recv_called
+
+    def set_name(self, name: str):
+        """
+        Set the name of the agent.
+        Why? because we need it to look like different agents
+        """
+        self._name = name
+
+    def _check_connection(self):
+        if self._agent_connector is None:
+            self._agent_connector = AutoGenConnector(self._network.lookup_agent(self.name))
+            self._terminate_connector = AutoGenConnector(self._network.lookup_termination())
+            time.sleep(0.01)
+
+    def receive(
+        self,
+        message: Union[Dict, str],
+        sender: Agent,
+        request_reply: Optional[bool] = None,
+        silent: Optional[bool] = False,
+    ):
+        """
+        Receive a message from the AutoGen system.
+        """
+        self._recv_called = True
+        self._check_connection()
+        self._agent_connector.send_receive_req(message, sender, request_reply, silent)
+
+    def generate_reply(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        exclude: Optional[List[Callable]] = None,
+    ) -> Union[str, Dict, None]:
+        """
+        Generate a reply message for the AutoGen system.
+        """
+        self._check_connection()
+        return self._agent_connector.send_gen_reply_req()
+
+    def _prepare_chat(self, recipient: "ConversableAgent", clear_history: bool, prepare_recipient: bool = True) -> None:
+        self._check_connection()
+        self._agent_connector.send_prep_chat(recipient, clear_history, prepare_recipient)
+
+    def send_terminate(self, recipient: "ConversableAgent") -> None:
+        self._check_connection()
+        self._agent_connector.send_terminate(recipient)
+        self._terminate_connector.send_terminate(self)
+        time.sleep(0.01)  # Let the network do things.
```

Comment:
Is there a way to do this that doesn't involve a sleep in library code?

Reply:
Removed sleep in everywhere except startup of broker.